### PR TITLE
Remtran cdb2api implementation (non-2pc, socksql mode)

### PR DESCRIPTION
**Pull request by dorinhogea**
PR switches remtran implementation to run over cdb2api instead of legacy remtran plugin.

Currently it supports non-2pc transactions and socksql transaction isolation. It provides protocol versioning and solves the stale schema change issue, which the previous implementation did not support.

PR adds a fdb compatibility testcase to check for ability to execute cross-cluster queries and writes with older and newer protocol versions.

The new implementation eliminates some of the fragile synchronization between read and write remote cursors in the previous implementation, by running the write statements remotely (push mode extended to writes).

### CODE REVIEW

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p>

File: db/fdb_push.c [hunk 1]
```diff
@@ -54,24 +56,51 @@ static int convert_policy_override_string_to_cdb2api_flag(char *policy) {
...
     GET_CLNT;
-    fdb_push_connector_t *push = NULL;
     struct Db *pDb = &pParse->db->aDb[node->remotedb];
 
     logmsg(LOGMSG_DEBUG,
-           "CALLING FDB PUSH RUN on query %s (gbl_fdb_push_remote=%d) (clnt force remote? %d) (clnt force redirect? "
+           "CALLING FDB PUSH RUN on query %s (fdb_push_remote=%d) (clnt force remote? %d) (clnt force redirect? "
            "%d) (clnt can redirect? %d)\n",
-           clnt->sql, gbl_fdb_push_remote, clnt->force_fdb_push_remote, clnt->force_fdb_push_redirect,
+           clnt->sql, clnt->fdb_push_remote, clnt->force_fdb_push_remote, clnt->force_fdb_push_redirect,
```
Comment (chands10) [inline db/fdb_push.c:101]: Why is  clnt->fdb_push_remote here twice?

File: db/fdb_push.c [hunk 2]
```diff
@@ -54,24 +56,51 @@ static int convert_policy_override_string_to_cdb2api_flag(char *policy) {
...
            clnt->can_redirect_fdb);
-    if (!gbl_fdb_push_remote && !clnt->force_fdb_push_remote && !clnt->force_fdb_push_redirect)
+    if (!clnt->fdb_push_remote && !clnt->force_fdb_push_remote && !clnt->force_fdb_push_redirect)
```
Comment (chands10) [inline db/fdb_push.c:103]: Here as well

File: db/db_tunables.h [hunk 3]
```diff
@@ -551,9 +551,10 @@ REGISTER_TUNABLE("foreign_db_allow_cross_class", NULL, TUNABLE_BOOLEAN, &gbl_fdb
 REGISTER_TUNABLE("foreign_db_resolve_local", NULL, TUNABLE_BOOLEAN, &gbl_fdb_resolve_local,
                  READONLY | NOARG | READEARLY, NULL, NULL, NULL, NULL);
 REGISTER_TUNABLE("foreign_db_push_remote", NULL, TUNABLE_BOOLEAN, &gbl_fdb_push_remote, NOARG, NULL, NULL, NULL, NULL);
+REGISTER_TUNABLE("foreign_db_push_remote_write", NULL, TUNABLE_BOOLEAN, &gbl_fdb_push_remote_write, NOARG, NULL, NULL, NULL, NULL);
 REGISTER_TUNABLE("foreign_db_push_redirect",
                  "Redirect fdb query to run via client instead of on server. (Default: off)", TUNABLE_BOOLEAN,
-                 &gbl_fdb_push_redirect_foreign, NOARG, NULL, NULL, NULL, NULL);
+                 &gbl_fdb_push_redirect_foreign, NOARG, NULL, NULL, fdb_push_write_update, NULL);
```
Comment (chands10) [inline db/db_tunables.h:557]: Is this supposed to be set on "foreign_db_push_remote" or "foreign_db_push_redirect"

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p>

File: db/db_tunables.h (hunk 3 above)
Comment (dorinhogea) [reply db/db_tunables.h:557]: right, foreign_db_push_remote_write

Comment (chands10) [reply db/db_tunables.h:557]: It looks like fdb_push_write_update is under the redirect tunable and not the write tunable

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/0 tests failed</b> &#x26A0;.</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>550/588 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>logfill [setup failure]<br>ixentry_order [setup failure]<br>sc_timepart_logicalsc_generated [setup failure]<br>sc_drop [setup failure]<br>random_osql_replay [setup failure]<br>queuedb_rollover_noroll1_generated [setup failure]<br>newsqlreplay [setup failure]<br>dohsql_race [setup failure]<br>analyze_fastinit_race [setup failure]<br>sc_repeated_updates [setup failure]</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>179/588 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>lostwrite [setup failure]<br>writes_remsql_names_rte_connect_generated<br>writes_remsql_names_partial_index_off_generated<br>writes_remsql_names_negativerowid_generated<br>writes_remsql_names<br>disttxn<br>snap_ha_retry_newsi_generated<br>sc_remsql<br>misstable_remsql<br>misstable_remsql_rte_connect_generated</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>9/588 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>lostwrite [setup failure]<br>sc_lock_inversion [setup failure]<br>disttxn<br>snapshot_during_truncate<br>mem_protobuf<br>snap_ha_retry_newsi_generated<br>simple_remsql_whitelist_generated<br>tunables<br>alias</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>7/588 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_tableversion_logicalsc_generated [setup failure]<br>sc_transactional_rowlocks_generated<br>phys_rep_tiered<br>snapshot_during_truncate<br>snap_ha_retry_newsi_generated<br>simple_remsql_whitelist_generated<br>auth</p>

**Review by mponomar — APPROVED** [review]

**Comment by dorinhogea** [comment]
@mponomar thank you

**Pull request merged**